PersistentStorage/Sources/Data/Repository/PersistentStorage+Codable/MainPersistentStorage+PersistentStorageCodable.swift

```diff
@@ -2,94 +2,66 @@ import Combine
-        return Just(())
-            .setFailureType(to: PersistentStorageError.self)
+                userDefaults.codableSet(
+                    value,
+                    forKey: key
+                )
+                // This operation is always successful
+                return ()
+            }
             .eraseToAnyPublisher()
     }
 
-    /// Method for reading a codable value with given key and type. Result is returned as a publisher.
-    ///
-    /// - Possible failures:
-    ///     - If type resolution with given parameter `valueType` fails, method will return `noValueFoundWithGivenType` failure.
-    ///     - If no value is associated with key given in parameter `valueKey` or its value is nil, the method will return `noValueFound` failure.
-    ///
-    /// - Parameters:
-    ///     - valueType: The type of value that you wish to read.
-    ///     - valueKey: Key that is assigned to a value that you wish to read.
-    public func readWithPublisher<T: Codable>(
+    public func readCodable<T: Codable>(
         valueType: T.Type,
-        valueKey: String
+        valueKey: String,
+        userDefaults type: UserDefaultsType
     ) -> AnyPublisher<T, PersistentStorageError> {
-        do {
-            let persistedValue = try getCodablePersistedValue(
-                valueType: valueType,
-                valueKey: valueKey
-            )
-
-            return Just(persistedValue)
-                .setFailureType(to: PersistentStorageError.self)
-                .eraseToAnyPublisher()
-        } catch {
-            guard let error = error as? PersistentStorageError else {
-                return Fail(error: PersistentStorageError.undefined)
-                    .eraseToAnyPublisher()
-            }
-
-            return Fail(error: error)
-                .eraseToAnyPublisher()
-        }
-    }
-
-    private func getCodablePersistedValue<T: Codable>(
-        valueType: T.Type,
-        valueKey: String
-    ) throws -> T {
-        guard let persistedValue: T = userDefaults.codableValue(forKey: valueKey) else {
-            if userDefaults.value(forKey: valueKey) != nil {
-                log(
-                    message: "❌ Value with given key exists, but method is unable to parse the value with given type.",
-                    for: .failure
+        userDefaults(for: type)
+            .flatMap { [weak self] userDefaults -> AnyPublisher<T, PersistentStorageError> in
+                guard let persistedValue: T = userDefaults.codableValue(forKey: valueKey) else {
+                    if userDefaults.value(forKey: valueKey) != nil {
+                        self?.log(
+                            message: "❌ Value with given key exists in userDefaults {\(type)}, but method is unable to parse the value with given codable type.",
+                            for: .failure
+                        )
+                        return Fail(error: PersistentStorageError.noValueFoundWithGivenType).eraseToAnyPublisher()
```

Comment:
No need to write enum name explicitly. Only `.noValueFoundWithGivenType` should be enough.